libr/bin/format/elf/elf.c

```diff
@@ -534,27 +539,44 @@ int Elf_(r_bin_elf_has_relro)(struct Elf_(r_bin_elf_obj_t) *bin) {
 	return 0;
 }
 
+/*
+To compute the base address, one determines the memory
+address associated with the lowest p_vaddr value for a
+PT_LOAD segment. One then btains the base address by
+truncating the memory address to the nearest multiple
+of the maximum page size
+*/
+
 ut64 Elf_(r_bin_elf_get_baddr)(struct Elf_(r_bin_elf_obj_t) *bin) {
 	int i;
-	/* hopefully.. the first PT_LOAD is base */
+	ut64 base = -1, tmp;
 	if (bin && bin->phdr) {
 		for (i = 0; i < bin->ehdr.e_phnum; i++) {
 			if (bin->phdr[i].p_type == PT_LOAD) {
-				return (ut64)bin->phdr[i].p_vaddr;
+				tmp = (ut64)bin->phdr[i].p_vaddr & ELF_PAGE_MASK;
+				if (tmp % (1 << ELF_PAGE_SIZE) != 0) {
+					tmp = tmp - (tmp % (1 << ELF_PAGE_SIZE));
+				}
+				if (tmp < base) base = tmp;
 			}
 		}
 	}
-	return 0;
+	return base == -1 ? 0 : base;
 }
 
 ut64 Elf_(r_bin_elf_get_boffset)(struct Elf_(r_bin_elf_obj_t) *bin) {
 	int i;
-	/* hopefully.. the first PT_LOAD is base */
+	ut64 base = -1, tmp;
 	if (bin && bin->phdr)
 		for (i = 0; i < bin->ehdr.e_phnum; i++)
-			if (bin->phdr[i].p_type == PT_LOAD)
-				return (ut64) bin->phdr[i].p_offset;
-	return 0;
+			if (bin->phdr[i].p_type == PT_LOAD) {
+				tmp =  (ut64)bin->phdr[i].p_offset & ELF_PAGE_MASK;
+				if (tmp % (1 << ELF_PAGE_SIZE) != 0) {
```

Comment:
You could avoid this condition and evaluate `tmp = tmp - (tmp % (1 << ELF_PAGE_SIZE));` all the time.